internal/layers/layers.go

```diff
@@ -154,7 +154,19 @@ func MountContainerLayers(ctx context.Context, layerFolders []string, guestRoot
 	if err != nil {
 		return "", fmt.Errorf("failed to add SCSI scratch VHD: %s", err)
 	}
-	containerScratchPathInUVM = scsiMount.UVMPath
+
+	// This handles the case where we want to share a containers scratch disk instead of mounting a
+	// new one. If the refcount is > 1 this means we've already mounted the scratch vhdx to the UVM.
+	// In OpenGCS they take the `ScratchPath` value in the CombinedLayers request and simply setup the
+	// upper and work directories for the overlay here. As we'd be passing in the exact same scratch
+	// space location this would run into trouble when going to setup a new overlay here as one already
+	// exists so we need to pass in a different value for `ScratchPath` that is still inside the mount point.
+	if scsiMount.RefCount() > 1 && uvm.OS() == "linux" {
+		scratchFmt := fmt.Sprintf("container_%s", filepath.Base(containerScratchPathInUVM))
+		containerScratchPathInUVM = ospath.Join("linux", scsiMount.UVMPath, scratchFmt)
```

Comment:
why the added "linux" at the beginning?

Reply:
It's not adding Linux to the beginning, it just uses the OS string passed in to decide which underlying `Join` function to call. https://github.com/microsoft/hcsshim/blob/c19ef4bd03a74f4ab40484963ad1d271a1820a0f/internal/ospath/join.go#L9